refactor(Merges Locations)

lomamech opened the pull request:
- Merges locations: Country, Province, Sector, Village
- Quick fixe in merge patient: prevent fatar error when no patient
  selected

closes #4599

lomamech committed:
refactor(Merges Locations)
- Merges locations: Country, Province, Sector, Village
- Quick fixe in merge patient: prevent fatar error when no patient
  selected

closes #4599

jmcameron commented on client/src/i18n/fr/form.json:
```diff
@@ -225,6 +225,7 @@
             "SAVE_SUCCESS": "Enregistrement avec succès",
             "SEARCH": "Recherche",
             "SET_EXCHANGE_CONVERSION": "Définir le taux de change pour la conversion",
+            "SELECT_LOCATION_CONSIDER": "Sélectionnez l'emplacement à considérer",
```
I am not good at French, but I noticed that the word  "Emplacement" is used here but "Localisations" is used  on line 278.    Is there a reason why the terminology is not consistent?   I have not looked further, but you agree this is an inconsistency, it would be worthwhile reviewing the rest of the modifications for this issue.

lomamech replied:
Thanks i will use "localisation"

jmcameron commented on client/src/modules/patients/registry/modals/mergePatients.modal.js:
```diff
@@ -27,6 +27,11 @@ function MergePatientsModalController(Patients, $state, Notify, Instance) {
   function submit() {
     if (vm.allAreEmployees) { return null; }
 
+    // Quick fix prevent fatar error : 2020-07-07 By: lomamech
```
Typo: fatar -> fatal

lomamech replied:
I realized that for the old module, merge of patiens, that when we do not select one and that validated, that abruptly stopped the server

lomamech replied:
My English is not correct because i use Google Translate 
 ![image](https://user-images.githubusercontent.com/8035588/86894089-7b3a5980-c0fa-11ea-9bdf-f7794f35872a.png)

jniles replied:
I always put my French into Google translate to get the accents correct 😆

jmcameron commented on server/controllers/admin/locations.js:
```diff
@@ -549,3 +549,73 @@ exports.delete.village = (req, res, next) => {
+      UPDATE province SET country_uuid = ? WHERE country_uuid = ?;`;
+
+    const removeOtherCountry = `
+      DELETE FROM country WHERE uuid IN (?);`;
```
Are you possibly deleting more than one country?   If we only delete one, perhaps this should use 'uuid = ?' rather than IN?

jmcameron commented on server/controllers/admin/locations.js:
```diff
@@ -549,3 +549,73 @@ exports.delete.village = (req, res, next) => {
+      .addQuery(replaceVillageInPatient1, [db.bid(selected), db.bid(other)])
+      .addQuery(replaceVillageInPatient2, [db.bid(selected), db.bid(other)])
+      .addQuery(removeOtherVillage, [db.bid(other)]);
+  }
```
I looked through this code for server 'merge()' while reviewing the SQL schema and it all looks okay to me (from a logic standpoint).

jmcameron commented on server/controllers/admin/locations.js:
```diff
@@ -549,3 +549,73 @@ exports.delete.village = (req, res, next) => {
+      UPDATE patient SET current_location_id = ? WHERE current_location_id = ?;`;
+
+    const replaceVillageInPatient2 = `
+      UPDATE patient SET origin_location_id = ? WHERE origin_location_id = ?;`;
```
This comment is probably outside the scope of this issue and pull request, but if a patient's current_location_id/origin_location_id is always a village, shouldn't 'village' be part of the name instead of 'location'?  Similarly for the 'location_id' in the other tables.  I seem to recall this discussion from 2015 and recall it was not an easy question to resolve and hard to generalize to different regions of the country (vs big cities, etc).

jniles replied:
Yes, that is correct, and I agree it is confusing.  It might be worth opening an issue over, and I've talked to @lomamech about doing so.

jmcameron reviewed (commented):
Another pretty light review.  I only scanned the client code but tried to look over the server 'merge()' code more closely.  

At some point in the future, I'd like to hear more about how the team balances making the front end client code prevent bad data from going to the server versus having the back end be bullet proof as well.

jniles reviewed (commented):
It looks like, if there are duplicate province names, merging the countries fails.

![image](https://user-images.githubusercontent.com/896472/86899373-d6bc1580-c101-11ea-93fc-653db61a43da.png)


Steps to reproduce:
  1. Make two countries, one named "DRC" and the other "RDC"
  2. Make a province in each country called "jniles"
  3. Try to merge "DRC" and "RDC"
 
The error is a unique key erreur since (country_uuid, name) must be unique in the provinces.

lomamech committed:
refactor(Merge location)
- Improve Requests use = uuid instead IN (uuid)
- Improve translation in french

lomamech committed:
Resolve conflict with branch Master

jniles commented on server/models/migrations/next/migrate.sql:
```diff
@@ -79,3 +79,15 @@ ALTER TABLE `service` DROP COLUMN `id`;
+
+ALTER TABLE `province` DROP INDEX `province_1`;
+
+ALTER TABLE `sector` DROP INDEX `sector_1`;
```
These needs to drop the key, not just the index.  I am still running into errors: 

![image](https://user-images.githubusercontent.com/896472/87141985-c2ac1b80-c29b-11ea-9828-df37e0962bba.png)

lomamech replied:
With the Vanga database, I was able to mix all the occurrences of DR Congo, and we can also see that there are even several Kwilu Province
![image](https://user-images.githubusercontent.com/8035588/87145885-fd18b700-c2a1-11ea-996c-c0e24c2e5fdb.png)

lomamech replied:
@jniles  Try to check in the structure of the province table, with the error message it seems to me that the index province_1 is not delete
![image](https://user-images.githubusercontent.com/8035588/87146231-8af4a200-c2a2-11ea-94b6-4242e879f400.png)

jniles commented on server/models/migrations/next/migrate.sql:
```diff
@@ -79,3 +79,15 @@ ALTER TABLE `service` DROP COLUMN `id`;
 
 
 ALTER TABLE `service` ADD PRIMARY KEY (`uuid`);
+  ('collectionCapacity', 'REPORT.COLLECTION_CAPACITY.TITLE');
```
I discovered what the issue was.  This migration script doesn't compile.

Can you remove the `collectionCapacity` part?  It breaks the script.

lomamech replied:
Okay

lomamech committed:
fix migrate file
